cohort-evaluator-spark/src/main/java/com/ibm/cohort/cql/terminology/HashedCode.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * (C) Copyright IBM Corp. 2021, 2021
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package com.ibm.cohort.cql.terminology;
+
+import java.util.Objects;
+
+import org.opencds.cqf.cql.engine.runtime.Code;
+
+
+/**
+ * This class is a simple wrapper class providing equals and hashCode functions
+ * for the org.opencds.cqf.cql.engine.runtime.Code class
+ * This is used to enable Set lookups for codes within ValueSets, which will be
+ * more performant than iterating through a list of codes
+ *
+ */
+public class HashedCode {
+	private Code code = null;
+
+	public HashedCode(Code code) {
+		this.code = code;
+	}
+
+	@Override
+	public boolean equals(Object o) {
+		if(o instanceof HashedCode) {
+			HashedCode hc = (HashedCode)o;
+			//we may get codes passed with with only the code, so compare the other fields only if they are present
+			if (code.getSystem() == null || (Objects.equals(hc.getCodeObject().getSystem(), code.getSystem())) &&
+				code.getVersion() == null || (Objects.equals(hc.getCodeObject().getVersion(), code.getVersion())) &&
+				code.getDisplay() == null || (Objects.equals(hc.getCodeObject().getDisplay(), code.getDisplay())) &&
+				code.getCode().equalsIgnoreCase(hc.getCodeObject().getCode())){
+				return true;
+			}
+		}
+		return false;
+	}
+	
+	public Code getCodeObject() {
+		return code;
+	}
+
+	@Override
+	public int hashCode() {
+		final int prime = 31;
+		int result = 1;
+		//only hash the code value since it is possible to get codes passed with
+		//without the other fields
```

Comment:
I believe that only hashing the code would allow for collisions if the same code existed in 2 different systems or with 2 different versions.  

If the value set contained ICD 250 v1 and ICD 250 v2, when loaded only one of these would be represented in the hashset.  If the comparison code only has the code we're fine, but it it also has the version and it's looking for v1 when only v2 is represented I think we have a problem.

I haven't written any test cases to prove this happens so happy if you can explain I'm wrong.

Reply:
There will be collisions but java handles collisions of 2 objects with the same hashCode by falling back to the equals() method. So objects with the same hashcode basically end up in a list and iterated over. I did add some tests to test against codes with different code system versions to verify, and did find a bug in that the valueSetImporter was not reading in the code system version from the spreadsheet, so we have a bug to fix here that I will address before I deliver.

Also, see my note below to Corey's comment on usage of the CodeKey class and why I implemented the equals method the way I did.

Reply:
Thanks!